llvm/tools/llvm-cgdata/llvm-cgdata.cpp

```diff
@@ -0,0 +1,352 @@
+                        ? sys::fs::OF_TextWithCRLF
+                        : sys::fs::OF_None);
+  if (EC)
+    exitWithErrorCode(EC, OutputFilename);
+
+  if (OutputFormat == CGDataFormat::Text) {
+    if (Error E = Writer.writeText(OS))
+      exitWithError(std::move(E));
+  } else {
+    if (Error E = Writer.write(OS))
+      exitWithError(std::move(E));
+  }
+
+  return 0;
+}
+
+static int show_main(int argc, const char *argv[]) {
+  std::error_code EC;
+  raw_fd_ostream OS(OutputFilename.data(), EC, sys::fs::OF_TextWithCRLF);
+  if (EC)
+    exitWithErrorCode(EC, OutputFilename);
+
+  auto FS = vfs::getRealFileSystem();
+  auto ReaderOrErr = CodeGenDataReader::create(Filename, *FS);
+  if (Error E = ReaderOrErr.takeError())
+    exitWithError(std::move(E), Filename);
+
+  auto Reader = ReaderOrErr->get();
+  if (ShowCGDataVersion)
+    OS << "Version: " << Reader->getVersion() << "\n";
+
+  if (Reader->hasOutlinedHashTree()) {
+    auto Tree = Reader->releaseOutlinedHashTree();
+    OS << "Outlined hash tree:\n";
+    OS << "  Total Node Count: " << Tree->size() << "\n";
+    OS << "  Terminal Node Count: " << Tree->size(/*GetTerminalCountOnly=*/true)
+       << "\n";
+    OS << "  Depth: " << Tree->depth() << "\n";
+  }
+
+  return 0;
+}
+
+static void parseArgs(int argc, char **argv) {
+  CGDataOptTable Tbl;
+  ToolName = argv[0];
+  llvm::BumpPtrAllocator A;
+  llvm::StringSaver Saver{A};
+  llvm::opt::InputArgList Args =
+      Tbl.parseArgs(argc, argv, OPT_UNKNOWN, Saver, [&](StringRef Msg) {
+        llvm::errs() << Msg << '\n';
+        std::exit(1);
+      });
+
+  if (Args.hasArg(OPT_help)) {
+    Tbl.printHelp(
+        llvm::outs(),
+        "llvm-cgdata <action> [options] (<binary files>|<.cgdata file>)",
+        ToolName.str().c_str());
+    std::exit(0);
+  }
+  if (Args.hasArg(OPT_version)) {
+    cl::PrintVersionMessage();
+    std::exit(0);
+  }
+
+  ShowCGDataVersion = Args.hasArg(OPT_cgdata_version);
+
+  auto parseFormat = [](const StringRef FT) {
+    return StringSwitch<CGDataFormat>(FT)
+        .Case("text", CGDataFormat::Text)
+        .Case("binary", CGDataFormat::Binary)
+        .Default(CGDataFormat::Invalid);
+  };
+  if (opt::Arg *A = Args.getLastArg(OPT_format)) {
+    StringRef OF = A->getValue();
+    OutputFormat = parseFormat(OF);
+    if (OutputFormat == CGDataFormat::Invalid)
+      exitWithError("unsupported format '" + OF + "'");
+  }
+
+  InputFilenames = Args.getAllArgValues(OPT_INPUT);
+  if (InputFilenames.empty())
+    exitWithError("No input file is specified.");
+  Filename = InputFilenames[0];
+
+  if (Args.hasArg(OPT_output)) {
+    OutputFilename = Args.getLastArgValue(OPT_output);
+    for (auto &Filename : InputFilenames)
+      if (Filename == OutputFilename)
+        exitWithError(
+            "Input file name cannot be the same as the output file name!\n");
+  }
+
+  SmallVector<opt::Arg *, 1> ActionArgs(Args.filtered(OPT_action_group));
```

Comment:
Do you expect this to ever have more than one element? Can we use `ArrayRef` or some constant-size container?

Reply:
Loop thru Args to ensure only one action arg, instead of using `SmallVector`.